### ISSUE TITLE
Allow reading journal without reading from disk.

### ISSUE THREAD

**Issue by psznm** [systemd/systemd#27140]
### Is your feature request related to a problem? Please describe

Unexpectedly, journalctl -n0 -f, reads from disk even though it should not need to.

When diagnosing system under heavy load with many GBs of logs, running journalctl will always make the situation worse as even journalctl -n0 -f may take many minutes and read several GBs if there is hundreds of GBs worth of logs. 

As expected, after the journalctl has finished the initial loading, it does not read anything from disk while still receiving new messages.

### Describe the solution you'd like

- New option that would prevent journalctl from reading from disk, instead it would only read from what is currently stored in memory, and/or new messages.
- Or just make the journalctl not read from disk when combining options -n0 -f

**Comment by vcaputo** [comment]
It's possible to improve the storage format such that less is faulted in to minimally begin "tailing the logs" though...

Also the way the kernel coarsely performs readahead/readaround generally prioritizes performance over minimizing IO/page-cache pressure.  It might make sense to suppress it in some tail-like journalctl invocations where it's more important to avoid hammering on storage and increasing memory pressure.

There used to be a knob in proc for controlling how much the kernel would read-around/ahead page faults in mmaps system-wide, but I don't seem to see it on my system anymore, maybe they got rid of it with the folio work?  When I last experimented with this it was noteworthy how little IO one could generate by shrinking the value drastically, while still being able to do the journal operations, albeit slower.  The journal objects are generally very small, and the access pattern random, so when the kernel tightens up its page fault handling it actually does result in much less reading w/journals.

**Comment by vcaputo** [comment]
> > If the reading causes IO then this means your system is under memory pressure and the memory that journald wrote the records into was so quickly reclaimed that the data had to be read back in when the readers read it.
> 
> I am sorry for my ignorance, but the records that I am requesting do not yet exist, and when I was testing it the system was not under any noteworthy memory pressure.

One case not covered by Lennart's comment is the plain old cold cache case, like when a system boots with plenty of preexisting journals and it's the first time you've queried them.  There's also the scenario of past memory pressure evicting once-cached journals from memory.  The kernel doesn't do anything more sophisticated than keeping the most recently used stuff in the cache, so a cache polluter can knock everything out and even a currently idle system may just have nothing presently useful sitting in the caches.

The root of your annoyance is arguably just unmet expectations.  You expect your command to behave like `tail -F /var/log/messages` where it just involves opening the file, seeking to the end, and waiting for new lines to be appended.  Unfortunately it's substantially more complicated than that with the journals as-is.

**Comment by psznm** [comment]
I see. My understanding of journald and journalctl is sadly full of assumptions. I only recently started digging into it because it became unusable in my situation. What I suggested was based on the assumption that `journalctl -f` first reads journal files, and then receives messages the same way journald does. Now that I know that is not true, let me rephrase my expectations and suggest something I imagine could be solutions.

As a system administrator, I need a reliable logging solution that will always work with me, and not against me when there is huge amounts of logs. Let me put that into context. The system I worked with had over 400G logs. With MaxFileSize set to 1G, it had around 1500 journal files. Running journalctl would take minutes of reading from disk as much as it could. Resulting in huge load on the server, reading gigabytes, to display lines that did not yet exist when it was ran. At that point journal is not a friend, but an enemy.

After some experimenting with the configuration, I started believing that is a misconfiguration, and that MaxFileSize might need to be much larger, and that would ensure much better performace, but since no such implications of the configuration option is mentioned in the manpages, it could also be seen as misbehaviour of journalctl.

Your last comment made me lose hope as I started believing that is simply how journald works and there is no way around it, but I have discovered --file option for journalctl. When I run `journalctl -f --file=/var/log/journal/*/system.journal`, on that system with over 400G logs it takes less than a second to start outputting lines. Compared to `journalctl -f` which seemingly achieves the same but takes minutes and gigabytes read from disk. My suggestion for journalctl not to read from disk, as this issue suggest, is not what I need. What I need is something like "_it takes less than a second to start outputting lines_"

This brings me to my ideas. I don't know the terminology, but `journalctl --disk-usage` talks about active and archived journals. I suppose active journals are those that are being written to, and archived are the ones that were rotated and we do not expect to ever again write to them.

Maybe running `journalctl -f` could start by checking only active journals and see if there is enough lines. If yes it could completely skip going through archived journals. If no, it would go through most recently archived journals until enough lines were found? This behaviour could also be combined with new option that would allow us to specify to how many journal files to limit the reading.

Or maybe journalctl could keep some sort of index file. Keeping metadata for each of the journal files. For example number of messages, first date, last date. This could make some operations like filtering by "until" and "since" extremely fast even with large amounts of files, as the journalctl would know which files to go through and skip the 99% that are not interesting for the query. This probably could/should come with options to reindex, or to not use the index at all.

**Comment by poettering** [comment]
journald doesn't scale well with huge number of log files. Increase your max file size
